### ISSUE TITLE
Compress blobs on disk?

### ISSUE THREAD

**Issue by michaelsproul** [sigp/lighthouse#5030]
## Description

I've think that blobs may compress up to ~50%, which would be a nice disk saving. Seeing as we're so close to Deneb on Goerli I think this will have to be a post-fork schema upgrade (v19).

## Steps to resolve

We'll need to change the signatures of `as_store_bytes` and `as_kv_store_op` to return `Result`, because compression is (usually) fallible. I've already done this on `tree-states` where lots of things are compressed with `zstd`:

https://github.com/sigp/lighthouse/blob/c8dc082ba8a385a4645e699466562ad5d9f97dd0/beacon_node/store/src/lib.rs#L326-L338

We could try to cherry-pick the changes from there, or just re-do them using mass find and replace. I think using `zstd` compression would also be good for consistency. It has very good performance both in terms of compression factor and speed.

We'll also need a schema upgrade v18 -> v19 that compresses all the blobs in the DB (we could try to shortcut this, but the UX would be worse).

**Comment by potuz** [comment]
are you sure about this? rollups will likely send already compressed data for blob transactions (Arbitrum already does this). If you had access to the original uncompressed txs that would make more sense. There may be consumers of uncompressed data blobs, but do we have numbers on these?

**Comment by michaelsproul** [comment]
@potuz I'm not sure actually, I thought I heard chatter on Discord about blob compressibility. If it's not useful it saves us some hassle

**Comment by michaelsproul** [comment]
Blobs are exactly 128KB and end up being padded with 0s if they aren't full. If users aren't using the full space and aren't sharing blobspace in an optimal way (requires clever packing), then aren't we wasting space by storing all these 0s?

Looking at the blobs on Sepolia at the moment, they compress by about 50% because of the 0s:

```
$ curl -s -H "Accept: application/octet-stream" "http://localhost:5052/eth/v1/beacon/blob_sidecars/0x65eed628f5fdd3455e5b8fdf2248a2e8e8df6ab56550aeb62fb79e7052b0a97f" > blobs.ssz
$ du -sh blobs.ssz
260K	blobs.ssz
$ zstd blobs.ssz 
blobs.ssz            : 50.09%   (   258 KiB =>    129 KiB, blobs.ssz.zst)
$ du -sh blobs.ssz.zst
132K	blobs.ssz.zst
```

**Comment by dapplion** [comment]
Makes sense to snappy compress blobs, since it's quite cheap and serving blobs is not a hot path. Mainnet usage patterns will probably be much more optimized than testnets yielding less compression, but shouldn't hurt.

**Comment by potuz** [comment]
I would suggest to first do the following experiment: grab a bunch of transactions (can be just arbitrary text for this part I think) and compress it with brotly (I know this is what Arbitrum does, but you can try whatever other rollups use) and then try compressing it with Snappy. I'm willing to bet a beer that a) you don't get gains most of the time and b) you obtain larger files often enough due to compression overhead. 

Edit: in reply to Michael, I've seen teams from different areas talking about a market for blob sharing and there's research about batch posting to maximize blob usage. It may be that there are gas conditions that would put zeroes in a blob, but I would definitely expect this to be the exception not the norm. As for packing, Arbitrum for sure (and surely others will do) will maximally pack using all bytes accounting for the difference between the field elements and the round number of bytes (but will not pack the last couple of bits that you would anyway miss with compression).

**Comment by michaelsproul** [comment]
Maybe worth waiting to see how packed blobs look on mainnet before gunning for this then.
